qa/rpc-tests/feature_zip239.py

```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021 The Zcash developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+from test_framework.mininode import (
+    NU5_PROTO_VERSION,
+    CInv,
+    NetworkThread,
+    NodeConn,
+    mininode_lock,
+    msg_getdata,
+    msg_mempool,
+    uint256_from_str,
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    BLOSSOM_BRANCH_ID,
+    HEARTWOOD_BRANCH_ID,
+    CANOPY_BRANCH_ID,
+    NU5_BRANCH_ID,
+    assert_equal,
+    assert_false,
+    assert_true,
+    fail,
+    hex_str_to_bytes,
+    nuparams,
+    p2p_port,
+    start_nodes,
+    wait_and_assert_operationid_status,
+)
+from tx_expiry_helper import TestNode
+
+import time
+
+# Test ZIP 239 behaviour before and after NU5.
+class Zip239Test(BitcoinTestFramework):
+    def setup_nodes(self):
+        return start_nodes(self.num_nodes, self.options.tmpdir, extra_args=[[
+            # Enable Canopy at height 205 to allow shielding Sprout funds first.
+            nuparams(BLOSSOM_BRANCH_ID, 205),
+            nuparams(HEARTWOOD_BRANCH_ID, 205),
+            nuparams(CANOPY_BRANCH_ID, 205),
+            nuparams(NU5_BRANCH_ID, 210),
+        ]] * self.num_nodes)
+
+    def cinv_for(self, txid, authDigest=None):
+        if authDigest is not None:
+            return CInv(5, txid, authDigest)
+        else:
+            return CInv(1, txid)
+
+    def verify_inv(self, testnode, txs):
+        # Make sure we are synced before sending the mempool message
+        testnode.sync_with_ping()
+
+        # Send p2p message "mempool" to receive contents from zcashd node in "inv" message
+        with mininode_lock:
+            testnode.last_inv = None
+            testnode.send_message(msg_mempool())
+
+        # Sync up with node after p2p messages delivered
+        testnode.sync_with_ping(waiting_for=lambda x: x.last_inv)
+
+        with mininode_lock:
+            msg = testnode.last_inv
+            assert_equal(len(msg.inv), len(txs))
+
+            expected_invs = sorted(txs, key=lambda inv: (inv.type, inv.hash, inv.hash_aux))
+            actual_invs = sorted(msg.inv, key=lambda inv: (inv.type, inv.hash, inv.hash_aux))
+
+            for (expected, actual) in zip(expected_invs, actual_invs):
+                fail_msg = "%r != %r" % (actual, expected)
+                assert_equal(actual.type, expected.type, fail_msg)
+                assert_equal(actual.hash, expected.hash, fail_msg)
+                assert_equal(actual.hash_aux, expected.hash_aux, fail_msg)
```

Comment:
I suggest implementing `__eq__` on [CInv](https://github.com/zcash/zcash/blob/f1896683b5b36edbc94a706a092410b61b2063b1/qa/rpc-tests/test_framework/mininode.py#L333-L355), so that this can just be `assert_equal(actual, expected)`.